### PR TITLE
feat: Add "Edit event" confirmation popup (M2-7188)

### PR DESCRIPTION
**Pull request by JustinNusca**
### 📝 Description

🔗 [M2-7188](https://mindlogger.atlassian.net/browse/M2-7188): [Admin Panel] Multi-informant - Add warning if admin has made changes in the 'Edit event' pop-up and closes it with not saving changes

This PR is a re-implementation of work originally done in an earlier task ([M2-6871](https://mindlogger.atlassian.net/browse/M2-6871)), which was undone as a result of addressing irreconcilable merge conflicts when merging [the multi-informant branch into develop](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1844).

There are similarities in the approach taken by the original work, but they are not 1:1. Some changes I made include:

- Updated the `ScheduleProvider` to include the `SaveChangesPopup` and to manage it's state. This was a context provider I previously added to wrap both the `Legend` and `Calendar` components in order to synchronize state changes across both of these components, and in particular to manage/render various popups that either component may open.
- Removed the `removeWarning` field from the `EventForm` component's form state. I did this so that the `dirty` state of the form could be correctly calculated, in order to determine if the "Save Changes" popup should be shown. `removeWarning` was not used as field, but as derived state for showing visible warning flags to the user, which are not reset when resetting changes. This variable is now passed via props to the relevant components (I also pluralized it, because it's an object that contains multiple possible warnings).
- Save a ref to the `EditEventForm`'s own ref object in the `ScheduleProvider`, and use that to determine if the `SaveChangesPopup` should be shown, as well as call the `EditEventForm`'s submission handler. This is a similar approach to the previous implementation, and `EditEventForm` already exposed a custom ref that used `useImperativeHandle` to pass along a few methods, including the submit handler. I'm not 100% happy with this, and it could be more conventionally done by lifting the form state up, but untangling this would have been a much more significant refactor.

### 📸 Screenshots

You may click each gif for a larger version.

| When Saving Changes | When Discarding Changes | When Cancelling "Save Changes" prompt | 
|-|-|-|
| when-saving-changes | when discarding changes | when cancelling |

### 🪤 Peer Testing

This flow should be replicable for both Scheduled and Unscheduled events, for both Default and Individual schedules.

For an applet with some activities…

1. Navigate to the **Applet → Schedule** and/or the **Participant Details → Schedule** screens.
2. Press an event in the calendar view to open the Edit Event popup.
3. Make a change to the event, and then attempt to close the popup without saving changes.
4. Notice that a "Save Changes" popup opens, prompting you to confirm or discard your changes.
    1. When pressing "Cancel" on the "Save Changes" popup:
      a. Notice that the "Save Changes" popup closes, and returns you to the "Edit Event" popup.
      b. After returning to the "Edit Event" popup, undo any changes you made to the event. 
      c. Close the popup without saving changes. Notice that the "Save Changes" popup does not open, because no changes have been recorded.
    2. When pressing "Don't Save" on the "Save Changes" popup:
        a. Notice that the "Save Changes" and "Edit Event" popups both close, and the event is _not_ updated.
    3. When pressing "Save" on the "Save Changes" popup:
        a. Notice that the "Save Changes" and "Edit Event" popups both close, and the event is updated.

[M2-7188]: https://mindlogger.atlassian.net/browse/M2-7188?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[M2-6871]: https://mindlogger.atlassian.net/browse/M2-6871?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/modules/Dashboard/features/Applet/Schedule/EventForm/EventForm.types.ts [hunk 1]
```diff
@@ -85,6 +87,7 @@ export type GetEventFromTabs = {
   hasNotificationsErrors?: boolean;
   hasAlwaysAvailableOption?: boolean;
   'data-testid'?: string;
+  removeWarnings?: Warning;
```
Comment (farmerpaul) [inline src/modules/Dashboard/features/Applet/Schedule/EventForm/EventForm.types.ts]: Just a suggestion to rename the `Warning` type as well, following on the heels of your `removeWarning` ⇒ `removeWarnings` change:
```suggestion
  removeWarnings?: EventFormWarnings;
```

**Review by farmerpaul — APPROVED** [review]
Looks great, with one small suggestion to rename one of the types.

Also, I noticed that when I change the Availability dropdown, and change it back to its original value, I'm still prompted to save changes. I suppose editing that value causes other form values to become dirty? No big deal IMO, I think it's fine to merge as is, unless it's an easy fix.

https://github.com/ChildMindInstitute/mindlogger-admin/assets/1282772/01b8c211-9ee4-4563-8aec-529c922197fb

**Review by mbanting — APPROVED** [review]
Thanks for re-applying this on the updated `develop` branch. Tested successfully locally and reviewed as well.

**Review by farmerpaul — APPROVED** [review]
Fix works great!

**Pull request merged**